File: mobile/lib/models/book.dart
class Book {
  final String id;
  final String title;
  final String author;
  final String description;
  final String coverUrl;
  final String audioUrl;
  final int duration; // in minutes
  final String category;
  final List<Chapter> chapters;
  final double rating;
  final int reviewCount;
  final bool isPremium;

  Book({
    required this.id,
    required this.title,
    required this.author,
    required this.description,
    required this.coverUrl,
    required this.audioUrl,
    required this.duration,
    required this.category,
    required this.chapters,
    this.rating = 0.0,
    this.reviewCount = 0,
    this.isPremium = false,
  });

  factory Book.fromJson(Map<String, dynamic> json) {
    return Book(
      id: json['id'].toString(),
      title: json['title'] ?? '',
      author: json['author'] ?? '',
      description: json['summary'] ?? json['description'] ?? '',
      coverUrl: json['cover_url'] ?? json['coverUrl'] ?? '',
      audioUrl: json['audio_url'] ?? json['audioUrl'] ?? '',
      duration: json['estimated_reading_minutes'] ?? json['duration'] ?? 0,
      category: json['genre'] ?? json['category'] ?? '',
      chapters: json['chapters'] != null
          ? (json['chapters'] as List)
                .map((chapter) => Chapter.fromJson(chapter))
                .toList()
          : [],
      rating: (json['rating_average'] ?? json['rating'] ?? 0).toDouble(),
      reviewCount: json['rating_count'] ?? json['reviewCount'] ?? 0,
      isPremium: json['is_premium'] ?? json['isPremium'] ?? false,
    );
  }

  factory Book.fromMap(Map<String, dynamic> map) => Book.fromJson(map);

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'author': author,
      'description': description,
      'coverUrl': coverUrl,
      'audioUrl': audioUrl,
      'duration': duration,
      'category': category,
      'chapters': chapters.map((chapter) => chapter.toJson()).toList(),
      'rating': rating,
      'reviewCount': reviewCount,
      'isPremium': isPremium,
    };
  }
}

class Chapter {
  final String id;
  final String title;
  final int duration; // in minutes
  final int startTime; // in seconds
  final int endTime; // in seconds

  Chapter({
    required this.id,
    required this.title,
    required this.duration,
    required this.startTime,
    required this.endTime,
  });

  factory Chapter.fromJson(Map<String, dynamic> json) {
    return Chapter(
      id: json['id'],
      title: json['title'],
      duration: json['duration'],
      startTime: json['startTime'],
      endTime: json['endTime'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'duration': duration,
      'startTime': startTime,
      'endTime': endTime,
    };
  }
}

class Quote {
  final String id;
  final String bookId;
  final String text;
  final int position;
  final String? chapterTitle;
  final DateTime createdAt;

  Quote({
    required this.id,
    required this.bookId,
    required this.text,
    required this.position,
    this.chapterTitle,
    required this.createdAt,
  });

  factory Quote.fromJson(Map<String, dynamic> json) {
    return Quote(
      id: json['id'],
      bookId: json['book_id'].toString(),
      text: json['text'],
      position: json['position'],
      chapterTitle: json['chapter_title'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  factory Quote.fromMap(Map<String, dynamic> map) => Quote.fromJson(map);

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'book_id': bookId,
      'text': text,
      'position': position,
      'chapter_title': chapterTitle,
      'created_at': createdAt.toIso8601String(),
    };
  }

  Map<String, dynamic> toMap() => toJson();
}

File: backend2/models/book.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// SortBy represents different sorting options for books
type SortBy string

const (
	SortByPopularity      SortBy = "popularity"      // download_count DESC
	SortByRating          SortBy = "rating"          // rating_average DESC
	SortByPublicationDate SortBy = "publication"    // created_at DESC
	SortByCreatedAt       SortBy = "created_at"      // created_at DESC  
	SortByTitle           SortBy = "title"           // title ASC
	SortByAuthor          SortBy = "author"          // author ASC
	SortByWordCount       SortBy = "word_count"      // word_count ASC
)

// Book represents a book from Aozora Bunko
type Book struct {
	ID                      int64     `json:"id" db:"id"`
	Title                   string    `json:"title" db:"title"`
	Author                  string    `json:"author" db:"author"`
	Epoch                   *string   `json:"epoch" db:"epoch"`
	WordCount               int       `json:"word_count" db:"word_count"`
	Embedding               []float64 `json:"embedding,omitempty" db:"embedding"`
	ContentURL              *string   `json:"content_url" db:"content_url"`
	Summary                 *string   `json:"summary" db:"summary"`
	Genre                   *string   `json:"genre" db:"genre"`
	DifficultyLevel         int       `json:"difficulty_level" db:"difficulty_level"`
	EstimatedReadingMinutes int       `json:"estimated_reading_minutes" db:"estimated_reading_minutes"`
	DownloadCount           int       `json:"download_count" db:"download_count"`
	RatingAverage           float64   `json:"rating_average" db:"rating_average"`
	RatingCount             int       `json:"rating_count" db:"rating_count"`
	IsPremium               bool      `json:"is_premium" db:"is_premium"`
	IsActive                bool      `json:"is_active" db:"is_active"`
	CreatedAt               time.Time `json:"created_at" db:"created_at"`
	UpdatedAt               time.Time `json:"updated_at" db:"updated_at"`
}

// Chapter represents a chapter within a book
type Chapter struct {
	ID        uuid.UUID `json:"id" db:"id"`
	BookID    int64     `json:"book_id" db:"book_id"`
	Title     string    `json:"title" db:"title"`
	Content   string    `json:"content" db:"content"`
	Position  int       `json:"position" db:"position"`
	WordCount int       `json:"word_count" db:"word_count"`
	CreatedAt time.Time `json:"created_at" db:"created_at"`
}

// Quote represents a quote from a book for recommendation system
type Quote struct {
	ID           uuid.UUID `json:"id" db:"id"`
	BookID       int64     `json:"book_id" db:"book_id"`
	Text         string    `json:"text" db:"text"`
	Position     int       `json:"position" db:"position"`
	ChapterTitle *string   `json:"chapter_title" db:"chapter_title"`
	CreatedAt    time.Time `json:"created_at" db:"created_at"`
}

// BookFilter represents filtering options for book queries
type BookFilter struct {
	Authors         []string `json:"authors,omitempty"`
	Genres          []string `json:"genres,omitempty"`
	Epochs          []string `json:"epochs,omitempty"`
	IsPremium       *bool    `json:"is_premium,omitempty"`
	MinWordCount    *int     `json:"min_word_count,omitempty"`
	MaxWordCount    *int     `json:"max_word_count,omitempty"`
	MinRating       *float64 `json:"min_rating,omitempty"`
	DifficultyLevel *int     `json:"difficulty_level,omitempty"`
	IsActive        *bool    `json:"is_active,omitempty"`
}

// BookSearchRequest represents a search request for books
type BookSearchRequest struct {
	Query  string      `json:"query,omitempty"`
	Filter *BookFilter `json:"filter,omitempty"`
	SortBy SortBy      `json:"sort_by,omitempty" validate:"omitempty,oneof=popularity rating publication title author word_count"`
	Limit  int         `json:"limit,omitempty" validate:"omitempty,min=1,max=100"`
	Offset int         `json:"offset,omitempty" validate:"omitempty,min=0"`
}

// BookListResponse represents a paginated list of books
type BookListResponse struct {
	Books   []*Book `json:"books"`
	Total   int     `json:"total"`
	Limit   int     `json:"limit"`
	Offset  int     `json:"offset"`
	HasMore bool    `json:"has_more"`
}

// CreateBookRequest represents the request body for creating a book
type CreateBookRequest struct {
	ID                      int64   `json:"id" validate:"required"`
	Title                   string  `json:"title" validate:"required,min=1,max=500"`
	Author                  string  `json:"author" validate:"required,min=1,max=200"`
	Epoch                   *string `json:"epoch" validate:"omitempty,max=50"`
	ContentURL              *string `json:"content_url" validate:"omitempty,url"`
	Summary                 *string `json:"summary" validate:"omitempty,max=2000"`
	Genre                   *string `json:"genre" validate:"omitempty,max=100"`
	DifficultyLevel         *int    `json:"difficulty_level" validate:"omitempty,min=1,max=5"`
	EstimatedReadingMinutes *int    `json:"estimated_reading_minutes" validate:"omitempty,min=0"`
	IsPremium               *bool   `json:"is_premium"`
}

// ToSQLOrderBy converts SortBy to SQL ORDER BY clause
func (s SortBy) ToSQLOrderBy() string {
	switch s {
	case SortByPopularity:
		return "download_count DESC, rating_average DESC"
	case SortByRating:
		return "rating_average DESC, rating_count DESC"
	case SortByPublicationDate:
		return "created_at DESC"
	case SortByTitle:
		return "title ASC"
	case SortByAuthor:
		return "author ASC, title ASC"
	case SortByWordCount:
		return "word_count ASC"
	default:
		return "download_count DESC, rating_average DESC"
	}
}